common/rpc/encryption/localStoreCertProvider.go

```diff
@@ -95,101 +128,142 @@ func (s *localStoreCertProvider) FetchServerRootCAsForClient(isWorker bool) (*x5
+func (s *localStoreCertProvider) FetchClientCertificate(isWorker bool) (*tls.Certificate, error) {
+
+	if s.tlsSettings == nil && s.workerTLSSettings == nil {
 		return nil, nil
 	}
+	certs, err := s.getCerts()
+	if err != nil {
+		return nil, err
+	}
+	if isWorker {
+		return certs.workerCert, nil
+	}
+	return certs.serverCert, nil
+}
+
+func (s *localStoreCertProvider) GetExpiringCerts(timeWindow time.Duration,
+) (CertExpirationMap, CertExpirationMap, error) {
+
+	expiring := make(CertExpirationMap)
+	expired := make(CertExpirationMap)
+	when := time.Now().UTC().Add(timeWindow)
+
+	certs, err := s.getCerts()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	checkError := checkTLSCertForExpiration(s.certs.serverCert, when, expiring, expired)
+	err = appendError(err, checkError)
+	checkError = checkTLSCertForExpiration(s.certs.clientCert, when, expiring, expired)
+	err = appendError(err, checkError)
+	checkError = checkTLSCertForExpiration(s.certs.workerCert, when, expiring, expired)
+	err = appendError(err, checkError)
+
+	checkCertsForExpiration(certs.clientCACerts, when, expiring, expired)
+	checkCertsForExpiration(certs.serverCACerts, when, expiring, expired)
+	checkCertsForExpiration(certs.serverCACertsWorker, when, expiring, expired)
+
+	return expiring, expired, err
+}
+
+func (s *localStoreCertProvider) getCerts() (*certCache, error) {
 
 	s.RLock()
-	if *cachedCertPool != nil {
+	if s.certs != nil {
 		defer s.RUnlock()
-		return *cachedCertPool, nil
+		return s.certs, nil
 	}
-
 	s.RUnlock()
 	s.Lock()
 	defer s.Unlock()
 
-	if *cachedCertPool != nil {
-		return *cachedCertPool, nil
+	if s.certs != nil {
+		return s.certs, nil
 	}
 
-	caPoolFromFiles, caCertsFromFiles, err := buildCAPoolFromFiles(files)
+	newCerts, err := s.loadCerts()
 	if err != nil {
 		return nil, err
 	}
+	s.certs = newCerts
+	return newCerts, nil
+}
 
-	caPoolFromData, caCertsFromData, err := buildCAPoolFromData(data)
-	if err != nil {
-		return nil, err
-	}
+func (s *localStoreCertProvider) loadCerts() (*certCache, error) {
 
-	if caPoolFromFiles != nil && caPoolFromData != nil {
-		return nil, errors.New(duplicateErrorMessage)
+	if s.tlsSettings == nil && s.workerTLSSettings == nil {
+		return nil, nil
 	}
 
-	if caPoolFromData != nil {
-		*cachedCertPool = caPoolFromData
-		*certs = caCertsFromData
-	} else {
-		*cachedCertPool = caPoolFromFiles
-		*certs = caCertsFromFiles
-	}
+	newCerts := certCache{}
+	var err error
 
-	return *cachedCertPool, nil
-}
+	if s.tlsSettings != nil {
+		newCerts.serverCert, err = s.fetchCertificate(s.tlsSettings.Server.CertFile, s.tlsSettings.Server.CertData,
+			s.tlsSettings.Server.KeyFile, s.tlsSettings.Server.KeyData)
+		if err != nil {
+			return nil, err
+		}
 
-func (s *localStoreCertProvider) FetchClientCertificate(isWorker bool) (*tls.Certificate, error) {
-	if isWorker {
-		return s.fetchWorkerCertificate()
-	}
-	if s.tlsSettings == nil {
-		return nil, nil
+		newCerts.clientCert = newCerts.serverCert
```

Comment:
Is there any situation where clientCert is not assigned the serverCert?

Reply:
Removed clientCert altogether.